nemo_aligner/metrics/common.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# A couple of common metrics as well as a handler class to provide a unified interface.
+
+from typing import Dict
+
+import hydra
+import torch
+from omegaconf import DictConfig
+from torchmetrics import Metric
+
+
+class InferenceMetricsHandler:
+    """A wrapper around metrics objects that will call update/compute/reset on all registered metrics.
+
+    If metrics_config is None, then all methods become no-ops and compute will return an empty dict.
+    """
+
+    def __init__(self, metrics_config: DictConfig):
+        if metrics_config is None:
+            metrics_config = {}
+        self.metrics = hydra.utils.instantiate(metrics_config)
+
+    def has_metrics(self) -> bool:
+        """Returns True if there are metrics to compute."""
+        return len(self.metrics) > 0
+
+    def update(self, batch: Dict, generation_output: Dict):
+        """Calling .update on all metrics.
+
+        Batch and generation output are coming directly from
+        validation dataloader and model.generate respectively.
+        """
+        for metric in self.metrics.values():
+            metric.update(batch, generation_output)
+
+    def compute(self) -> Dict[str, float]:
+        """Returns a dictionary with finalized metric values."""
+        return {name: metric.compute() for name, metric in self.metrics.items()}
+
+    def reset(self):
+        """Will reset state of all metrics to prepare for the next validation run."""
+        for metric in self.metrics.values():
+            metric.reset()
+
+
+def get_prediction(idx, batch, generation_output):
+    """A helper function to get predicted output for a typical dataloader format."""
+    elem_metadata = batch["metadata"][idx]
+    # need to remove original prompt from the generated output as it's always included
+    if "conversations" in elem_metadata:  # chat format
+        # last one is the target, so splitting on the one before that
+        pred_output = generation_output["sentences"][idx].split(elem_metadata["conversations"][-2]["value"])[-1]
```

Comment:
Hard to follow, can you break it down in several steps?

Reply:
yeah, this logic is messy and might not even be correct. Let me explain it better in the code, but maybe we can try to find a better way to handle this - basically we need to get the output from the model, but it always includes prompt as well as the final output. And I was not able to come up with a clean way of how to strip out the prompt in a robust manner (seems that there is token length, but they are not directly mappable to the word positions). Let me know if you have any suggestions here!

Reply:
I see, thanks for the detailed explanation in the code. I agree that it's a recurring pain point in NeMo's generate() that it's tricky to isolate the Assistant response.

I'm not sure how the code flows right now, but in the chat format it'd probably be more reliable if you could keep track of the prompt that was fed to the model (which must be created at some point, right before it's tokenized), so that you could use a similar logic as in the input-output format here. The concern with your current implementation is that `last_prompt_part` might be present in the Assistant response (imagine for instance that the user just asked "why" without a question mark).

Minor remark: you can use `rsplit(..., 1)` to do a single split when you care only about the last component.